crates/ruff_textwrap/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+//! Functions related to adding and removing indentation from lines of
+//! text.
+
+use std::borrow::Cow;
+use std::cmp;
+
+use ruff_newlines::StrExt;
+
+/// Indent each line by the given prefix.
+///
+/// # Examples
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("First line.\nSecond line.\n", "  "),
+///            "  First line.\n  Second line.\n");
+/// ```
+///
+/// When indenting, trailing whitespace is stripped from the prefix.
+/// This means that empty lines remain empty afterwards:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("First line.\n\n\nSecond line.\n", "  "),
+///            "  First line.\n\n\n  Second line.\n");
+/// ```
+///
+/// Notice how `"\n\n\n"` remained as `"\n\n\n"`.
+///
+/// This feature is useful when you want to indent text and have a
+/// space between your prefix and the text. In this case, you _don't_
+/// want a trailing space on empty lines:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent("foo = 123\n\nprint(foo)\n", "# "),
+///            "# foo = 123\n#\n# print(foo)\n");
+/// ```
+///
+/// Notice how `"\n\n"` became `"\n#\n"` instead of `"\n# \n"` which
+/// would have trailing whitespace.
+///
+/// Leading and trailing whitespace coming from the text itself is
+/// kept unchanged:
+///
+/// ```
+/// use ruff_textwrap::indent;
+///
+/// assert_eq!(indent(" \t  Foo   ", "->"), "-> \t  Foo   ");
+/// ```
+pub fn indent<'a>(text: &'a str, prefix: &str) -> Cow<'a, str> {
+    if prefix.is_empty() {
+        return Cow::Borrowed(text);
+    }
+
+    let mut result = String::with_capacity(2 * text.len());
+    let trimmed_prefix = prefix.trim_end();
+    for line in text.universal_newlines() {
+        if line.trim().is_empty() {
+            result.push_str(trimmed_prefix);
+        } else {
+            result.push_str(prefix);
+        }
+        result.push_str(line.as_full_str());
+    }
+    Cow::Owned(result)
+}
+
+/// Removes common leading whitespace from each line.
+///
+/// This function will look at each non-empty line and determine the
+/// maximum amount of whitespace that can be removed from all lines:
+///
+/// ```
+/// use ruff_textwrap::dedent;
+///
+/// assert_eq!(dedent("
+///     1st line
+///       2nd line
+///     3rd line
+/// "), "
+/// 1st line
+///   2nd line
+/// 3rd line
+/// ");
+/// ```
+pub fn dedent(text: &str) -> Cow<'_, str> {
+    // Find the minimum amount of leading whitespace on each line.
+    let mut prefix = None;
+    for line in text.universal_newlines() {
+        if let Some((index, _)) = line.char_indices().find(|(_, ch)| !ch.is_whitespace()) {
+            prefix = Some(cmp::min(prefix.unwrap_or(index), index));
+        }
+    }
+    let Some(prefix) = prefix else {
+        return Cow::Borrowed(text);
+    };
```

Comment:
What do you think of initializing `prefix` with `usize::MAX` as marker for 'no prefix'. Any non-empty string will change the value to a lower value